qat graph mode: add combine_conv_bn pass

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39828 [wip] qat graph mode: add combine_conv_bn pass**
* #39837 eager qat: refactor conv-bn to make conv with scaled weights a func
* #39840 qat graph mode: make `_forward` of per channel observers scriptable
* #39796 qat graph mode: initial scaffolding for prepare_qat_script

Summary:

Adds the JIT pass to combine conv and bn before training.  Note: this only adds the MVP,
there are various TODOs for future PRs to handle various edge cases.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

vkuzo committed:
[wip] qat graph mode: add combine_conv_bn pass

Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/qat_combine_conv_bn.cpp:
```diff
@@ -0,0 +1,315 @@
+  // Dot in the ".Conv2d" and ".BatchNorm2d" is an attempt to
+  // prevent matching module's whose name might end with Conv2d
+  // But are user defined modules.
+  const PatternInfo pattern = PatternInfo::parse_from_str(R"IR(
+graph(%self, %x):
+    %conv_submodule = match::module[name=".Conv2d"](%self)
+    %conv_out = prim::CallMethod[name="forward"](%conv_submodule, %x)
+    %bn_submodule = match::module[name=".BatchNorm2d"](%self)
+    %bn_out = prim::CallMethod[name="forward"](%bn_submodule, %conv_out)
+    return (%bn_out))IR");
```
btw we have different way of matching now: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/fold_conv_bn.cpp.html#358, this might help with expressing the transformation with subgraph_rewriter as well

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] qat graph mode: add combine_conv_bn pass"


Summary:

not ready for review yet.

Adds the JIT pass to combine conv and bn before training.

Test Plan:

PYTORCH_JIT_LOG_LEVEL=">qat_combine_conv_bn" python test/test_quantization.py TestQuantizeQATScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21987858](https://our.internmc.facebook.com/intern/diff/D21987858)

[ghstack-poisoned]

vkuzo commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -79,27 +79,33 @@ def freeze_bn_stats(self):
 
+    # Note: this should be kept in sync with
+    # torch/csrc/jit/passes/quantization/qat_combine_conv_bn.cpp
     def _forward(self, input):
```
the changes to this function do not change the logic, they just make it easier to inspect and reason about the resulting IR and write the equivalent IR in the JIT pass.

vkuzo commented on torch/quantization/quantize_jit.py:
```diff
@@ -57,7 +57,9 @@ def _prepare_jit(model, qconfig_dict, inplace=False, quant_type=QuantType.STATIC
+        model = wrap_cpp_module(torch._C._jit_pass_qat_combine_convbn(model._c))
+    else:
         model = fuse_conv_bn_jit(model, inplace)
     model_c = torch._C._jit_pass_insert_observers(model._c,
```
necessary changes here will come in a future PR

jerryzh168 commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2959,25 +2959,50 @@ def test_single_linear_dynamic(self):
                 super(M, self).__init__()
-                self.conv = torch.nn.Conv2d(1, 1, 1)
-                self.bn = torch.nn.BatchNorm2d(1)
+                self.conv11 = torch.nn.Conv2d(1, 1, 1)
```
why do we change the name?

jerryzh168 commented on torch/csrc/jit/passes/quantization/qat_combine_conv_bn.cpp:
```diff
@@ -0,0 +1,90 @@
+  void transform(Module& module);
+};
+
+std::pair<std::string, std::string> getBeforeAndAfterPatterns() {
```
would be better to put the patterns and filters together

jerryzh168 replied:
[optional] nit: rename this function to `getQATConvBnPattern` or something more informative

vkuzo replied:
sure

jerryzh168 commented on tools/build_variables.bzl:
```diff
@@ -199,6 +199,7 @@ core_sources_full = [
     "torch/csrc/jit/passes/utils/subgraph_utils.cpp",
     "torch/csrc/jit/passes/xnnpack_rewrite.cpp",
     "torch/csrc/jit/passes/vulkan_rewrite.cpp",
+    "torch/csrc/jit/passes/quantization/qat_combine_conv_bn.cpp",
```
nit: suggest to use `qat_fold_conv_bn` to be aligned with the inference version

vkuzo replied:
we aren't actually folding the bn in this pass though, this pass is fundamentally doing a different thing compared to the `fold_conv_bn` pass

jerryzh168 commented on torch/csrc/jit/passes/quantization/qat_combine_conv_bn.cpp:
```diff
@@ -0,0 +1,90 @@
+namespace jit {
+
+namespace {
+using graph_rewrite_helper::PatternInfo;
```
this is not used

jerryzh168 commented on torch/csrc/jit/passes/quantization/qat_combine_conv_bn.cpp:
```diff
@@ -0,0 +1,90 @@
+  SubgraphRewriter rewriter;
+  rewriter.RegisterRewritePattern(patterns.first, patterns.second);
+
+  for (auto& method : module.get_methods()) {
```
this does not handle child modules right now. will need to add that, example: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/insert_observers.cpp.html#1108

vkuzo replied:
yep, sg.  I have a TODO in the test case to handle that in a future PR

jerryzh168 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -79,27 +79,33 @@ def freeze_bn_stats(self):
-            torch.qat_conv2d_and_unscale(input, scaled_weight, scale_factor,
-                                         self.bias, self.stride, padding,
-                                         self.dilation, self.groups)
+            conv_res = \
+                torch.qat_conv2d_and_unscale(input, scaled_weight, scale_factor,
+                                             self.bias, self.stride, _pair(0),
+                                             self.dilation, self.groups)
```
why do we repeat this function call?

vkuzo replied:
good point, I'll add a comment.  It has to do with making this module scriptable, I'll have to get the actual error if we don't do this.  It's good for this to be scriptable in order to be able to compare the generated IR to the JIT pass.
